tabulator/stream.py

```diff
@@ -452,30 +459,34 @@ def skip_negative_rows(extended_rows):
             Rows to skip are taken from  Stream.__skip_rows_by_numbers
             """
             rows_to_skip = [n for n in self.__skip_rows_by_numbers if n < 0]
-            if not rows_to_skip:
-                for row in extended_rows:
+            buffer_size = abs(min(rows_to_skip))
+            # collections.deque - takes O[1] time to push/pop values from any side.
+            buffer = deque()
+
+            # use buffer to save last rows
+            for row in extended_rows:
+                buffer.append(row)
+                if len(buffer) > buffer_size:
+                    yield buffer.popleft()
+
+            # Now squeeze out the buffer
+            global last_row_number
```

Comment:
why use the global? usually it's a sign of bad design.
my conclusion is that you should run the 'remove rows from end' bit  _before_ the original 'remove_rows' processor - and then this entire thing is not needed.

Reply:
oh my stupid head :) that's so obvious )))